projects/plutus/src/ltc_afe.c

```diff
@@ -0,0 +1,308 @@
+#include "ltc_afe.h"
+#include "ltc68041.h"
+
+static bool s_discharging_cells[LTC_CELLS_PER_DEVICE * LTC_AFE_DEVICES_IN_CHAIN] = { false };
+
+static uint16_t s_read_reg_cmd[NUM_LTC_AFE_REGISTER] = {
+  LTC6804_RDCFG_RESERVED,
+  LTC6804_RDCVA_RESERVED,
+  LTC6804_RDCVB_RESERVED,
+  LTC6804_RDCVC_RESERVED,
+  LTC6804_RDCVD_RESERVED,
+  LTC6804_RDAUXA_RESERVED,
+  LTC6804_RDAUXA_RESERVED,
+  LTC6804_RDSTATA_RESERVED,
+  LTC6804_RDSTATB_RESERVED,
+  LTC6804_RDCOMM_RESERVED
+};
+
+static void prv_wakeup_idle(const LTCAFESettings *afe) {
+  gpio_set_pin_state(&afe->cs, GPIO_STATE_LOW);
+  delay_us(2);
+  gpio_set_pin_state(&afe->cs, GPIO_STATE_HIGH);
+}
+
+static StatusCode prv_read_register(const LTCAFESettings *afe,
+                                    LTCAFERegister reg, uint8_t *data, size_t len) {
+  if (reg > NUM_LTC_AFE_REGISTER) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  uint16_t reg_cmd = s_read_reg_cmd[reg];
+  uint8_t cmd[4] = { 0 };
+
+  cmd[0] = (uint8_t)(reg_cmd >> 8);
+  cmd[1] = (uint8_t)(reg_cmd & 0xFF);
+  uint16_t cmd_pec = crc15_calculate(cmd, 2);
+
+  cmd[2] = (uint8_t)(cmd_pec >> 8);
+  cmd[3] = (uint8_t)(cmd_pec & 0xFF);
+
+  return spi_exchange(afe->spi_port, cmd, 4, data, len);
+}
+
+static StatusCode prv_read_voltage(LTCAFESettings *afe, LTCAFEVoltageRegister reg, uint8_t *data) {
+  if (reg > NUM_LTC_AFE_VOLTAGE_REGISTER) {
+    return status_code(STATUS_CODE_INVALID_ARGS);
+  }
+
+  // 6 bytes in register + 2 bytes for PEC
+  size_t len = ((LTC_AFE_CELLS_IN_REG * 2) + 2) * LTC_AFE_DEVICES_IN_CHAIN;
+
+  if (reg == LTC_AFE_VOLTAGE_REGISTER_A) {
+    return prv_read_register(afe, LTC_AFE_REGISTER_CELL_VOLTAGE_A, data, len);
+  }
+  if (reg == LTC_AFE_VOLTAGE_REGISTER_B) {
+    return prv_read_register(afe, LTC_AFE_REGISTER_CELL_VOLTAGE_B, data, len);
+  }
+  if (reg == LTC_AFE_VOLTAGE_REGISTER_C) {
+    return prv_read_register(afe, LTC_AFE_REGISTER_CELL_VOLTAGE_C, data, len);
+  }
+  return prv_read_register(afe, LTC_AFE_REGISTER_CELL_VOLTAGE_D, data, len);
+}
+
+// start cell voltage conversion
+static void prv_trigger_adc_conversion(const LTCAFESettings *afe) {
+  uint8_t mode = (uint8_t)((afe->adc_mode + 1) % 3);
+  // ADCV command
+  uint16_t adcv = LTC6804_ADCV_RESERVED | LTC6804_ADCV_DISCHARGE_PERMITTED
+                  | LTC6804_CNVT_CELL_ALL | (mode << 7);
+  uint8_t cmd[4] = { 0 };
+
+  cmd[0] = (uint8_t)(adcv >> 8);
+  cmd[1] = (uint8_t)(adcv & 0xFF);
+
+  uint16_t cmd_pec = crc15_calculate(cmd, 2);
+  cmd[2] = (uint8_t)(cmd_pec >> 8);
+  cmd[3] = (uint8_t)(cmd_pec);
+
+  prv_wakeup_idle(afe);
+
+  spi_exchange(afe->spi_port, cmd, 4, NULL, 0);
+
+  // wait for conversions to finish
+  delay_ms(100);
+}
+
+static void prv_trigger_aux_adc_conversion(const LTCAFESettings *afe) {
+  uint8_t mode = (uint8_t)((afe->adc_mode + 1) % 3);
+  // ADAX
+  uint16_t adax = LTC6804_ADAX_RESERVED | LTC6804_ADAX_GPIO1 | (mode << 7);
+  uint8_t cmd[4] = { 0 };
+  cmd[0] = (uint8_t)(adax >> 8);
```

Comment:
Honestly not really a fan of unions because of the lack of portability. I don't think the boilerplate is too terrible, especially since it's just a mix of grabbing the high bits and lower bits.

I could maybe write a macro that splits a ``uint16_t`` into high bits and low bits to make it explicit what's going on?